### DISCUSSION TITLE
Question in terms of ICEORYX ERROR  when publishing data. (chunk and images)

### DISCUSSION THREAD

**Discussion by GeYapeng** [eclipse-iceoryx/iceoryx#856] [Q&A] [answered]
The error occured when publishing  images in the type of 'uint8_t [rows][cols]' in a topic. (Sometimes it might succeed  while the warning information still existed but mostly it failed.)
I'll appreciate it if anyone can help with this or just offer some advice.
-------------

The program was running  in the method below in a docker container.
Terminal-2 failed after a few seconds. It  stopped with 'Segmentation fault'. 
-------------
       Terminal-1  ~#   iox-roudi

        [Warning]Reserving 60128824 bytes in the shared memory [/iceoryx_mgmt]
        [ Reserving shared memory successful ] 
        Reserving 694525840 bytes in the shared memory [/root]
        [ Reserving shared memory successful ] 
        RouDi is ready for clients

        [Warning]: Application iox_pub not responding (last response 1544 milliseconds ago) --> removing it
-------------
       Terminal-2(pub)  ~#  sh run.sh

       Mempool [m_chunkSize = 524328, numberOfChunks = 50, used_chunks = 50 ] has no more space left
       [ Error ]: MemoryManager: unable to acquire a chunk with a chunk-payload size of 481044The following mempools are available: 
       MemPool [ ChunkSize = 168, ChunkPayloadSize = 128, ChunkCount = 10000 ]  
       MemPool [ ChunkSize = 1064, ChunkPayloadSize = 1024, ChunkCount = 5000 ]  
       MemPool [ ChunkSize = 16424, ChunkPayloadSize = 16384, ChunkCount = 1000 ]  
       MemPool [ ChunkSize = 131112, ChunkPayloadSize = 131072, ChunkCount = 200 ]  
       MemPool [ ChunkSize = 524328, ChunkPayloadSize = 524288, ChunkCount = 50 ] 
       MemPool [ ChunkSize = 1048616, ChunkPayloadSize = 1048576, ChunkCount = 30 ] 
       MemPool [ ChunkSize = 4194344, ChunkPayloadSize = 4194304, ChunkCount = 20 ] 
       MemPool [ ChunkSize = 8388648, ChunkPayloadSize = 8388608, ChunkCount = 20 ]  
       MemPool [ ChunkSize = 33554472, ChunkPayloadSize = 33554432, ChunkCount = 10 ]
       [Warning]: ICEORYX error! MEPOO__MEMPOOL_GETCHUNK_POOL_IS_RUNNING_OUT_OF_CHUNKS
       Unable to loan data, error code: 1

       Segmentation fault

The program of publishing the images:
--------------
topic_data.hpp:

       #define ROWS 480
       #define COLS  752
       struct Data
       {
          uint8_t pub_img[ROWS][COLS];
       }

iox_pub.cpp:

       auto pub = publisher.loan();
       auto& pub_data = pub.value();

       extern cv::Mat img;    // This is the img to publish.
       if (!img.empty())
       {
            for (int i = 0; i < ROWS; i++)
            {
                for (int j = 0; j < COLS; j++)
                {
                     pub_data->pub_img[i][j] = img.at<uint8_t>(i, j);
                }
           }
       }

iox_sub.cpp:

       auto sub = subscriber.take();
       auto& sub_data = sub.value();

       cv::Mat sub_img(ROWS, COLS, CV_8UC1);   // Can  'CV_8UC1' be replaced by 'CV_8UC3' or other forms? 
       for (int i = 0; i < ROWS; i++)
       {
            for (int j = 0; j < COLS; j++)
            {
                     sub_img.at<uint8_t>(i,j) = sub_data->pub_img[i][j];
            }
       }

**Comment by elfenpiff** [comment, marked as answer]
@GeYapeng There seems to be two problems, the roudi is running out of chunks since there are too many samples in use and the publisher does not handle the case when roudi is unable to provide samples. When roudi is running out of chunks  `publisher.loan()` will return an invalid sample since there aren't any left and this case seems to be not covered.

1. To fix the segmentation fault in the publisher (handle running out of chunks) adjust the code like this:
```cpp
publisher.loan()
    .and_then([&](auto& pub_data) {
        extern cv::Mat img;  // This is the img to publish.
        if (!img.empty()) {
            for (int i = 0; i < ROWS; i++) {
                for (int j = 0; j < COLS; j++) {
                    pub_data->pub_img[i][j] = img.at<uint8_t>(i, j);
                }
            }
        }
    })
    .or_else([](auto) {
        std::cout << "Mempool is running out of chunks" << std::endl;
    });
```

Here you can find further use cases on how to publish samples: https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_examples/icedelivery/iox_publisher.cpp

If you prefer the classical `if-then-else` approach you can also look at this example: https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_examples/icehello/iox_publisher_helloworld.cpp


2. If you require more chunks of a certain size then our default memory config is offering you can always configure your own memory pool. Take a look at: https://iceoryx.io/v1.0.1/advanced/configuration-guide/

I think you can reuse the existing mempool configuration and just increase the number of chunks for the sample size 524328 to 100. This is what in the end the error message states right before the application crashed with a segmentation fault:
```
   Terminal-2(pub)  ~#  sh run.sh

   Mempool [m_chunkSize = 524328, numberOfChunks = 50, used_chunks = 50 ] has no more space left
   [ Error ]: MemoryManager: unable to acquire a chunk with a chunk-payload size of 481044The following mempools are available:
```

Here is our default mempool config with the adjusted sample size, just follow the configuration guide and use it. 
```
[general]
version = 1

[[segment]]

[[segment.mempool]]
size = 128
count = 10000

[[segment.mempool]]
size = 1024
count = 5000

[[segment.mempool]]
size = 16384
count = 1000

[[segment.mempool]]
size = 131072
count = 200

[[segment.mempool]]
size = 524288
count = 100

[[segment.mempool]]
size = 1048576
count = 30

[[segment.mempool]]
size = 4194304
count = 20

[[segment.mempool]]
size = 8388608
count = 20

[[segment.mempool]]
size = 33554432
count = 10
```

**Reply by elBoberido** [reply]
@elfenpiff I think we still have a `std::terminate` in case we are running out of chunks. The upper layer would support it, but in the lower layer it's not yet implemented

@GeYapeng in addition to what elfenpiff said, on subscriber side you are accessing the sample without checking if it's valid. See also https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/how-optional-and-error-values-are-returned-in-iceoryx.md#expected.

Your code should look like this
```cpp
   auto sub = subscriber.take();
   if (!sub.has_error())
   {
       auto& sub_data = sub.value();

       cv::Mat sub_img(ROWS, COLS, CV_8UC1);   // Can  'CV_8UC1' be replaced by 'CV_8UC3' or other forms? 
       for (int i = 0; i < ROWS; i++)
       {
            for (int j = 0; j < COLS; j++)
           {
                    sub_img.at<uint8_t>(i,j) = sub_data->pub_img[i][j];
           }
      }
   }
```
or with the more functional approach
```cpp
   subscriber.take()
             .and_then([$] (auto& data) {
      cv::Mat ...
   });
```

I guess the reason you are running out of chunks is that by default there are only 100 chunks of the size you need. I guess you also used the default `SubscriberOptions` which has a larger value than 100 for `queueCapacity`. Please have a look at https://github.com/eclipse-iceoryx/iceoryx/tree/master/iceoryx_examples/iceoptions on how to use the options.
It might also be worth to look at the waitset and callback example if you don't want to do polling.

And finally, since I don't have the full code, do you have any sleeps in your code?

**Reply by GeYapeng** [reply]
@elfenpiff @elBoberido
It's really my pleasure to receive a reply.  I'll give it a go with the advice immediately.
And I added sleeps in the loop :
 
    while (!killswitch) 
    {
         ...  ...
          std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }

**Reply by GeYapeng** [reply]
I added a  `roudi.toml` setting file as suggested and then run  `RouDi` with it . It worked in chunk size warning but the segmentation fault still existed . So I checked other parts of my code, especially the part where the image is loaded or resized, and finally found that it was the copy method of the image that lead to ' Segmentation fault  ' . Using `copyTo()` seems to be much safer than `clone()` in OpenCV  before I `extern` the image . 

The suggestions do have helped me a lot and the problem is solved  now.
Thank you very much ! ! ! @elfenpiff  @elBoberido
